### PR TITLE
Added PubSubMessage attributes support for Google Cloud PubSub.

### PR DESCRIPTION
**Pull request by vladimir-malyk**
Issues #510 #738

### CODE REVIEW

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/Model.scala [hunk 1]
```diff
@@ -4,10 +4,15 @@
...
+import java.time.Instant
+
 import scala.collection.immutable
 import scala.collection.JavaConverters._
 
-final case class PubSubMessage(data: String, messageId: String)
+final case class PubSubMessage(data: String,
+                               messageId: String,
+                               attributes: Option[immutable.Map[String, String]] = None,
+                               publishTime: Option[Instant] = None)
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/Model.scala]: Could you add a Java-friendly API with a backwards compatible contructor, `withAttributes(java.util.Map)` and `withPublishTime(Instant)` methods, please.

**Comment by ennru** [comment]
@vladimir-malyk Any chance you get to do the improved Java API? We're closing in for the next release.

**Comment by vladimir-malyk** [comment]
> Any chance you get to do the improved Java API? We're closing in for the next release.

@ennru unfortunately, I am not familiar with Java-friendly API good-practices, based on Scala.

For now, I have no enough spare time to find out issue in details on my own. 

Could you advice such a materials for fast reading? Or just write, what I have to do exactly.

ps. as far as I see from my perspective, there are no way to provide backward compatible constructor without affecting scala api.

**Comment by ennru** [comment]
It is not that much that needs to be added to get a Java API here.
We've collected a bit of advice in [contributor advice](https://github.com/akka/alpakka/blob/master/contributor-advice.md).
You may for example have a look at `MqttConnectionSettings` in the MQTT connector.

File: google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/Model.scala [hunk 2]
```diff
@@ -4,10 +4,23 @@
...
+                               attributes: Option[immutable.Map[String, String]] = None,
+                               publishTime: Option[Instant] = None) {
+
+  def withAttributes(attributes: java.util.Map[String, String]): PubSubMessage =
+    copy(attributes = Some(attributes.asScala.toMap))
+
+  def withPublishTime(publishTime: Instant): PubSubMessage =
+    copy(publishTime = Some(publishTime))
+
+}
```
Comment (ennru) [inline google-cloud-pub-sub/src/main/scala/akka/stream/alpakka/googlecloud/pubsub/Model.scala]: We'd need a companion object with a Java-friendly `create` method to get by without Scala Option in the interface:
`def create(data: String, messageId: String): PubSubMessage`

File: google-cloud-pub-sub/src/test/java/akka/stream/alpakka/googlecloud/pubsub/ExampleUsageJava.java [hunk 3]
```diff
@@ -56,7 +57,7 @@ private static void example() throws NoSuchAlgorithmException, InvalidKeySpecExc
 
         //#publish-single
         PubSubMessage publishMessage =
-                new PubSubMessage("1", new String(Base64.getEncoder().encode("Hello Google!".getBytes())));
+                new PubSubMessage("1", new String(Base64.getEncoder().encode("Hello Google!".getBytes())), Option.apply(null), Option.apply(null));
```
Comment (ennru) [inline google-cloud-pub-sub/src/test/java/akka/stream/alpakka/googlecloud/pubsub/ExampleUsageJava.java]: ...and use `create` instead of `new` here.

**Comment by ennru** [comment]
Thank you for your contribution! Keep them coming.

**Pull request merged**